src/hp/derp.rs

```diff
@@ -186,10 +188,33 @@ async fn write_frame(
     for b in bytes {
         writer.write_all(b).await?;
     }
-    writer.flush().await?;
     Ok(())
 }
 
+/// AsyncWrites a complete frame, errors if it is unable to write within the given `timeout`.
+/// Ignores the timeout if `timeout.is_zero()`
+///
+/// Does not flush.
+async fn write_frame_timeout(
+    writer: impl AsyncWrite + Unpin,
+    frame_type: FrameType,
+    bytes: Vec<&[u8]>,
+    timeout: Duration,
+) -> Result<()> {
+    if timeout.is_zero() {
+        return write_frame(writer, frame_type, bytes).await;
+    }
+    tokio::select! {
```

Comment:
[`tokio::time::timeout`](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html) is a lot simpler to use than a manual `select`